### PR TITLE
[#10481] [feat] Add Fuzzy Speculative Decoding (FSD) support

### LINKED ISSUE #10481: [Feature]: Adding support for Fuzzy Speculative Decoding (ACL Findings 2025)
### 🚀 The feature, motivation and pitch

Hi everyone -- I’d like to propose adding support for Fuzzy Speculative Decoding (FSD), a small extension to speculative decoding that provides an **optional, tunable quality–throughput tradeoff** by relaxing strict distributional equivalence when accepting draft tokens. We recently published FSD at ACL 2025 and are happy to provide a PR that implements FSD (with all required tests, benchmarks, etc.), but first wanted to check first whether such a PR would be in-scope and would potentially be accepted.

**What is FSD:** Production deployments often need to dynamically control the latency-quality tradeoff of their generations in response to shifting constraints -- for instance to hit different cost targets across traffic tiers or to adapt to changing hardware / queue conditions. While standard speculative decoding (SD) delivers substantial inference speedups, its gains are effectively binary: since SD strictly accept draft tokens that fully preserve the target model's distribution, it offers no mechanism to intentionally trade a small, controlled amount of quality for materially higher throughput. **FSD addresses this limitation by introducing a user-controlled threshold T that modulates draft-token acceptance** based on how close target and draft next-token distributions are (i.e., whether the divergence between these distributions is below the threshold). This gives users a simple knob to directly control the draft-token acceptance rate, enabling smooth quality–throughput tuning.

**Impact on generation quality:** Our results show that FSD provides a smooth, tunable quality–throughput curve. At lower threshold settings, FSD can match SD benchmark accuracy while already delivering noticeably higher throughput (commonly ~10–20% faster). As the threshold increases, FSD yields substantially larger speedups (~30–50% over SD) in exchange for small, controlled quality reductions (typically within ~2% relative degradation on standard benchmarks reported in the paper).

**Proposed implementation plan:** Given its simplicity, FSD can be implemented to:
- Require minimal changes to the current speculative decoding implementation (simply adding fully optional FSD acceptance logic to the SD rejection path in `acceptDraftTokensKernel`)
- Be gated and opt-in (default disabled), leaving standard speculative decoding completely untouched unless explicitly enabled
- Be cheap when activated, introducing minimal additional computation only when draft tokens are rejected by standard SD, while leaving the dominant acceptance path unchanged
- Apply to any draft-target model pair out-of-the-box (i.e., fully training-free)

Concretely, I’m proposing we implement the “reducible” variant described in the paper: when FSD is enabled, keep the existing SD acceptance unchanged, and only if a draft token is rejected do an additional “fuzzy” check.
1. Run normal SD acceptance for candidate token x_i
2. If SD accepts → accept and move to next token (exact current behavior)
3. If SD rejects and fsd_enabled:
    - Compute a divergence or distance between targetProbs and draftProbs (e.g., KL divergence, or another lightweight metric)
    - If Div < T → override rejection and accept
    - Else → keep rejection and proceed with standard resampling behavior

**Would FSD be considered in-scope for TensorRT-LLM’s speculative decoding support?** If so, I'll submit a PR implementing the reducible FSD variant behind an opt-in flag, along with unit tests, accuracy benchmarks, and throughput measurements across representative models.

Thank you!

### Alternatives

There are currently no alternatives to FSD that facilitate a smooth generation latency-quality tradeoff for any draft-target model pair. 

### Additional context

Our ACL 2025 Findings paper introducing FSD can be found [HERE](https://github.com/user-attachments/files/24465007/2025.findings-acl.1346.pdf). You can also find our Hugging Face `custom_generate` implementation of FSD at [maxholsman/fuzzy-spec-dec](https://huggingface.co/maxholsman/fuzzy-spec-dec), which lets users run FSD directly in `transformers` with any draft-target model pair available through on HF. 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and checked the [documentation](https://nvidia.github.io/TensorRT-LLM/) and [examples](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples) for answers to frequently asked questions.

### PR DESCRIPTION
**Pull request by maxholsman**
## Description

Closes #10481.

Implements **Fuzzy Speculative Decoding (FSD)** as described in [our ACL 2025 Findings paper](https://github.com/user-attachments/files/24465007/2025.findings-acl.1346.pdf). The implementation is minimal, fully opt-in, and leaves standard speculative decoding completely untouched unless explicitly enabled.

**Note:** The original paper reports using JS divergence as the FSD divergence used to collect the main results. We've since discovered a minor bug in the calculation of JS divergence in our experiment code. The empirical results in the paper therefore correspond to `reverse_kl`, as implemented below.

### What this PR does

Standard speculative decoding (SD) accepts draft tokens only when they preserve the target model's distribution exactly (via a rejection-sampling test). This is statistically correct, but gives users no way to trade a small, controlled amount of quality for meaningfully higher throughput.

FSD adds a single tunable threshold `T` to the rejection path. When SD rejects a draft token, FSD checks whether the divergence between the target and draft next-token distributions is below `T`. If so, the token is accepted anyway ("fuzzy acceptance"); if not, standard resampling proceeds as usual. Setting `T = 0` recovers exact SD. Increasing `T` smoothly raises the acceptance rate and throughput in exchange for a small, tunable quality reduction.

**The acceptance path is completely unchanged.** FSD only adds a lightweight divergence check on tokens that SD has already rejected, so it introduces zero overhead on accepted tokens and minimal overhead overall.

### Implementation

All changes are contained to the external-draft-tokens speculative decoding path:

**C++ kernel** (`cpp/tensorrt_llm/kernels/speculativeDecoding/externalDraftTokensKernels.cu`)
- When SD rejects a token and `fsdThreshold > 0`, the kernel computes a divergence between `targetProbs` and `draftProbs` across the full vocabulary using a block-level parallel reduction.
- If `divergence < fsdThreshold`, the rejection is overridden and the token is accepted.
- Four divergence metrics are supported (selectable via `fsdDivergenceType`):
  - `js` / `0` — Jensen-Shannon divergence, normalized by ln(2) to [0, 1]. **Recommended:** bounded and interpretable; `T = 1.0` accepts everything, `T = 0.5` is a natural midpoint.
  - `kl` / `1` — KL(draft ‖ target): large when the draft is overconfident on tokens the target discounts.
  - `reverse_kl` / `3` — KL(target ‖ draft): large when the target has high-probability modes the draft missed entirely.
  - `tv` / `2` — Total variation distance, in [0, 1].

**C++ executor API** (`cpp/include/tensorrt_llm/executor/executor.h`)
- `ExternalDraftTokensConfig` gains two new optional fields:
  - `fsdThreshold` (`std::optional<float>`) — the acceptance threshold `T`. Disabled when absent or 0.
  - `fsdDivergenceType` (`std::optional<int32_t>`) — selects the divergence metric (0–3).

**Python API** (`tensorrt_llm/runtime/model_runner_cpp.py`)
- `ModelRunnerCpp.generate()` accepts two new per-request keyword arguments:
  - `fsd_threshold` — scalar or per-request list of floats.
  - `fsd_divergence_type` — scalar or per-request list; accepts integers 0–3 or string names (`"js"`, `"kl"`, `"tv"`, `"reverse_kl"`).

**Python bindings** (`cpp/tensorrt_llm/nanobind/executor/request.cpp`)
- Exposes `fsd_threshold` and `fsd_divergence_type` as read-only properties on `ExternalDraftTokensConfig`.

**Bug fix: `skipDecoding` aliasing** — fixed a pre-existing bug in `acceptDraftTokensKernel` where `finishedInput` and `finishedOutput` could alias the same buffer. The old code wrote the skip-decoding flag and then read it back via the aliased pointer, causing the FSD override to read a stale value. Fixed by capturing `finishedInput[batchSlot]` into a local variable before any writes.

### Usage

```python
from tensorrt_llm.runtime import ModelRunnerCpp

runner = ModelRunnerCpp.from_dir(engine_dir="...", ...)

outputs = runner.generate(
    batch_input_ids=input_ids,
    draft_tokens_list=draft_tokens,
    draft_logits_list=draft_logits,   # raw (pre-softmax) logits from the draft model
    max_new_tokens=200,
    fsd_threshold=1.0,                # T=1.0 accepts all tokens with JS < 1.0 (i.e., everything)
    fsd_divergence_type="js",         # or "kl", "reverse_kl", "tv", or an integer 0–3
)
```

`fsd_threshold` and `fsd_divergence_type` can also be passed as per-request lists when batching requests with different settings.

**Recommended starting point:** `fsd_divergence_type="js"` with `fsd_threshold` swept over `[0.2, 0.4, 0.6, 0.8, 1.0]`. Since JS is normalized to [0, 1], the threshold has a stable, model-independent interpretation across draft-target pairs.

### Results

Benchmarked on Llama 3.2 1B (draft) + Llama 3.2 3B (target), `draft_len=5`, A5000 GPU:

| Divergence | Threshold | Tok/s | Accept% | Tok/round | vs SD |
|---|---|---|---|---|---|
| SD baseline | — | 269 | 77% | 14.1 | — |
| JS | 0.6 | 335 | 83% | 15.2 | +25% |
| JS | 0.8 | 385 | 92% | 17.7 | +43% |
| JS | **1.0** | **438** | **100%** | **20.0** | **+63%** |

JS saturates exactly at `T = 1.0` as expected from the [0, 1] normalization. At `T = 0.8` (~92% acceptance), throughput is already +43% vs SD with only ~2% quality reduction on standard benchmarks (see paper for full accuracy results).

Benchmarked on Llama 3.1 8B (draft) + Llama 3.1 70B (target), `draft_len=5`, 4×A6000 (`tp=4`):

| Divergence | Threshold | Tok/s | Accept% | Tok/round | vs SD |
|---|---|---|---|---|---|
| SD baseline | — | 15.4 | 80% | 3.98 | — |
| reverse_kl | 1.0 | 16.6 | 83% | 4.15 | +8% |
| reverse_kl | 2.0 | 19.2 | 92% | 4.62 | +25% |
| reverse_kl | 3.0 | 20.1 | 97% | 4.87 | +31% |
| reverse_kl | **4.0** | **20.3** | **100%** | **4.98** | **+32%** |

reverse_kl saturates at T ≈ 4 for open-ended generation (vs T ≈ 0.5–1.1 in the paper's structured benchmarks — see note above). The relative speedup at saturation is consistent: +31–34% across both settings.

## Test Coverage

- **`cpp/tests/unit_tests/layers/externalDraftTokensLayerTest.cpp`** — three new `TYPED_TEST` cases exercising the FSD CUDA kernel directly through the `ExternalDraftTokensLayer`:
  - `FsdDisabledMatchesStandardSD`: verifies T=0 produces identical output to standard SD.
  - `FsdHighThresholdAcceptsAllRejections`: verifies T=100 fuzzy-accepts every SD-rejected token, with correct output token ids at each step.
  - `FsdPartialThresholdSelectiveAcceptance`: verifies T=0.5 (JS) selectively fuzzy-accepts only low-divergence rejections (JS≈0.40) while leaving high-divergence rejections (JS≈1.0) to standard SD correction.
- **`tests/unittest/bindings/test_executor_bindings.py`** — verifies `fsd_threshold` and `fsd_divergence_type` round-trip correctly through the `ExternalDraftTokensConfig` nanobind bindings.
- **`tests/unittest/test_model_runner_cpp_fsd_utils.py`** — unit tests for `_map_fsd_divergence_type` (string/int normalization, error cases) and `_normalize_fsd_param` (scalar/list broadcasting).
- **`examples/draft_target_model/run_dtm_fsd_test.py`** — end-to-end integration script validating FSD acceptance rates and throughput against a real draft+target model pair.

## PR Checklist

Please review the following before submitting your PR:
- [x] PR description clearly explains what and why.
- [x] PR follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of our knowledge.
- [x] Test cases are provided for new code paths (kernel-level C++ tests, Python binding tests, Python utility tests).
- [x] No new dependencies have been introduced.
- [x] [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) — no ownership changes; no update required.
- [x] Documentation updated: usage example added to PR description; example scripts in `examples/draft_target_model/`.
- [x] tava architecture diagram — no significant design change; no update required.
- [x] Reviewers assigned automatically are appropriate for this PR.

- [ ] Please check this after reviewing the above items as appropriate for this PR.

## Release Notes

* **New Features**
  * Added Fuzzy Speculative Decoding (FSD) with configurable thresholds and divergence metrics (JS, KL, TV, Reverse KL)
  * Extended external draft tokens configuration with FSD parameters
  * Added Python API support for FSD parameters in speculative decoding

* **Documentation**
  * Added example scripts demonstrating FSD usage and configuration

* **Tests**
  * Added unit tests for FSD functionality and parameter validation

### CODE REVIEW

**Comment by karljang** [comment]
Just an update, we’re discussing the future direction in the original #10481.
